platforms/nuttx/src/px4/common/px4_crypto.cpp

```diff
@@ -0,0 +1,138 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#if defined(PX4_CRYPTO)
+
+#include <px4_crypto.h>
+#include <px4_platform_common/arch_crypto.h>
+
+extern "C" {
+#include <nuttx/random.h>
+}
+
+/*
+ * For now, this is just a dummy up/down counter for tracking open/close calls
+ */
+static int crypto_open_count = 0;
+
+void px4_crypto_init()
+{
+	// Initialize nuttx random pool, if it is being used by crypto
+#ifdef CONFIG_CRYPTO_RANDOM_POOL
+	up_randompool_initialize();
+#endif
+
+	// initialize keystore functionality
+	arch_keystore_init();
+
+	// initialize actual crypto algoritms
+	arch_crypto_init();
+}
+
+bool px4_crypto_open(px4_crypto_session_handle_t &handle, px4_crypto_algorithm_t algorithm)
+{
+	handle.lock();
+
+	// Check that it is not already open
+	if (handle.valid()) {
+		handle.unlock();
+		return false;
+	}
+
+	// Open the HW specific arch handle
+	handle.arch_handle = arch_crypto_open(algorithm);
+
+	if (!arch_crypto_handle_valid(handle.arch_handle)) {
+		handle.unlock();
+		return false;
+	}
+
+	handle.handle = ++crypto_open_count;
+	handle.unlock();
+
+	return true;
+}
+
+void px4_crypto_close(px4_crypto_session_handle_t &handle)
+{
+	handle.lock();
+
+	// Check that it is not already closed
+	if (!handle.valid()) {
+		handle.unlock();
+		return;
+	}
+
+	crypto_open_count--;
+	handle.handle = 0;
+	handle.unlock();
+
+	arch_crypto_close(&(handle.arch_handle));
+}
+
+
+bool px4_encrypt_data(px4_crypto_session_handle_t &handle,
+		      uint8_t  key_index,
+		      const uint8_t *message,
+		      size_t message_size,
+		      uint8_t *cipher,
+		      size_t *cipher_size)
+{
+	return arch_encrypt_data(handle.arch_handle, key_index, message, message_size, cipher, cipher_size);
```

Comment:
Do we actually need to wrap these?

Reply:
That is a good question; I was originally thinking that we'd have a HW/platform specific layer and an OS-specific layer separately. Another reason for this was, that I wanted to have the "backend" as C-interface (to be usable e.g. with bootloader) and then the actual interface to the px4 modules as C++ (which came out only half-way).

Reply:
Ok I don't mind much, but unless there's good reason, having a common API, also used by the BL, to me is preferable. And then do a C++ API around that. Or considering directly using C++ for the BL.

Reply:
Ok, for now I kept the C-coded "backend" and implemented a proper C++ wrapper class (actual interface) meant for generic use. This split also centralizes the usage into px4_layer, which gives (me :) an opportunity to implement the OS-specific call gate for memory protected build which is also in my agenda... And the bootloader can use the C-coded backend directly.